File: binsifter/core/av_detect.py
"""Antivirus product detection - a follow-up to defender.py's exclusion
button, since Windows Defender is only one of many antivirus products an
analyst's machine might be running. The existing exclusion button only
ever helps if Defender is the active product; this module answers the
"which AV is actually installed here" question first, so Settings can
point the analyst at the right place even when it isn't Defender.

Two independent detection mechanisms, one per supported platform - Windows
and Linux have no shared registry of "what antivirus is installed," so
this isn't one implementation with a platform branch, it's two genuinely
different approaches:

**Windows** uses the same real-world mechanism Windows' own Security app
is built on: the `root/SecurityCenter2` WMI namespace's `AntiVirusProduct`
class, which every registered antivirus product (Defender included)
publishes itself into - this is how Windows Security's "Virus & threat
protection" page itself knows what's installed, not a BinSifter
invention. Read via a plain `Get-CimInstance` call in a non-elevated
PowerShell subprocess, same shell-out approach defender.py already uses
(no `wmi`/`pywin32` dependency needed) - unlike the actual exclusion
change, reading this WMI class does NOT require Administrator rights on
any supported Windows version.

Two real caveats worth stating plainly about the Windows path:
- `root/SecurityCenter2` is a client-SKU feature (Windows 10/11 desktop) -
  the underlying Security Center service does not exist on Windows Server,
  so this returns an empty list there rather than raising, and Settings
  shows a message explaining why instead of implying nothing is installed.
- Windows can suppress Defender's own entry from this class once a
  third-party real-time AV registers as active (Defender drops to
  "passive mode" but keeps running signature updates) - so "Defender not
  in this list" does not necessarily mean Defender isn't present at all,
  only that something else is the active real-time scanner. This module
  reports what SecurityCenter2 itself says is active; it does not try to
  second-guess that.

**Linux** has no equivalent centralized registry at all - there is no
Linux analog to SecurityCenter2 that every AV/EDR vendor publishes itself
into. Instead this checks three independent, non-privileged signals per
known product (a systemd unit file, a running process name under /proc,
and a representative install path) against a curated table
(`_LINUX_AV_SIGNATURES`) built from each vendor's own Linux agent
documentation, and reports a hit if ANY signal matches. This is
deliberately NOT exhaustive - a product not on that table, or one
installed under a renamed/repackaged service, simply won't be found, and
an empty result here means "nothing on the known list was detected," a
weaker guarantee than Windows' "SecurityCenter2 says nothing is
registered." Both are still reported as a normal empty list, not an
error, since neither is a detection failure.

Deliberately does NOT attempt to script an exclusion for anything but
Windows Defender (defender.py's job) - most other vendors either have no
documented, stable local CLI for this at all, or (for enterprise EDR
products especially - CrowdStrike, SentinelOne, Symantec/Broadcom
Endpoint Protection managed via SEP Manager) deliberately block local
self-exclusion by design, since letting a local process silently exempt
itself from EDR scanning is exactly the kind of thing EDR exists to
prevent. `guidance_for()` instead returns a short, honest pointer to where
that vendor's own exclusion settings usually live - checked against
Linux-specific phrasing first (`_LINUX_VENDOR_GUIDANCE`) before falling
back to the original Windows-console-oriented table, since the same
vendor's Linux agent very often exposes an entirely different exclusion
mechanism (a config file or CLI flag instead of a GUI settings page).
"""

from __future__ import annotations

import json
import logging
import platform
import subprocess
from dataclasses import dataclass
from pathlib import Path

logger = logging.getLogger(__name__)


class AvDetectionError(Exception):
    """Raised only for hard failures - on Windows: powershell.exe missing,
    or the query itself timing out. On any platform other than Windows or
    Linux entirely (macOS, etc.) - detection simply isn't implemented
    there yet. A genuinely empty result (no AV registered/detected) on
    either Windows or Linux is NOT an error, it's a valid `[]` return, so
    callers can show it as a plain status message instead of an error
    dialog.
    """


@dataclass
class AvProduct:
    name: str


# Substring (lowercased) -> short pointer to that vendor's own exclusion
# settings. Deliberately not step-by-step (menu paths shift between
# versions) - just enough to tell the analyst where to look instead of
# guessing, plus an honest note where local self-exclusion typically isn't
# possible at all (centrally-managed EDR).
_VENDOR_GUIDANCE: tuple[tuple[str, str], ...] = (
    ("windows defender", "Use the button below - BinSifter can add this exclusion automatically."),
    ("microsoft defender", "Use the button below - BinSifter can add this exclusion automatically."),
    ("mcafee", "McAfee: Endpoint Security console > Threat Prevention > Exclusions (or pushed from ePO if centrally managed)."),
    ("eset", "ESET: open the product > Setup > Detection Engine (or Antivirus) > Exclusions."),
    ("sophos", "Sophos: Sophos Endpoint > Exclusions, or Sophos Central > policy Exclusions if centrally managed."),
    ("symantec", "Symantec/Broadcom Endpoint Protection: exclusions are normally pushed from the SEP Manager console - a local client usually can't add its own."),
    ("broadcom", "Broadcom Endpoint Protection: exclusions are normally pushed from the SEP Manager console - a local client usually can't add its own."),
    ("crowdstrike", "CrowdStrike Falcon: exclusions are managed centrally from the Falcon console by design - there is no supported local self-exclusion."),
    ("sentinelone", "SentinelOne: exclusions are managed centrally from the Management Console by design - there is no supported local self-exclusion."),
    ("bitdefender", "Bitdefender: Protection > Antivirus > Settings (gear icon) > Manage Exceptions."),
    ("kaspersky", "Kaspersky: Settings > Threats and Exclusions > Manage Exclusions."),
    ("trend micro", "Trend Micro: open the console/agent's Scan Exclusion List settings."),
    ("malwarebytes", "Malwarebytes: Settings > Exclusions > Add Exclusion."),
    ("avast", "Avast: Menu > Settings > General > Exceptions."),
    ("avg", "AVG: Menu > Settings > General > Exceptions."),
    ("webroot", "Webroot: PC Security > Identity & Privacy Shields > Application/Exclusion list."),
    ("f-secure", "F-Secure: Settings > find the exclusion/exception list for real-time scanning."),
    ("norton", "Norton: Settings > Antivirus > Scans and Risks > Exclusions/Low Risks."),
)

# Linux-specific phrasing, checked BEFORE _VENDOR_GUIDANCE above - the same
# vendor's Linux agent frequently exposes a completely different exclusion
# mechanism (a config file or CLI flag) than its Windows GUI, so reusing
# the Windows-console wording here would actively mislead a Linux analyst.
# Only vendors with a real, documented Linux agent get an entry - anything
# not listed here still falls through to the generic Windows-flavored
# table below rather than making something up.
_LINUX_VENDOR_GUIDANCE: tuple[tuple[str, str], ...] = (
    ("clamav", "ClamAV: add an Exclude/ExcludePath directive to /etc/clamav/clamd.conf (and freshclam.conf if relevant), then restart the clamav-daemon service."),
    ("defender for endpoint", "Microsoft Defender for Endpoint on Linux: run `mdatp exclusion path add --path <folder>`, or set it via your organization's managed configuration profile if enrolled - the button below only automates Windows Defender, not the Linux agent."),
    ("sophos", "Sophos for Linux: check your Sophos Central policy first if centrally managed (local overrides are often blocked by design); otherwise exclusions are set via savsetup/the on-box config."),
    ("bitdefender", "Bitdefender GravityZone (Linux): exclusions are normally pushed from the GravityZone Control Center console, not set locally on the agent."),
    ("trend micro", "Trend Micro Deep Security Agent: exclusions are normally configured from the Deep Security Manager console, not set locally on the agent."),
    ("symantec", "Symantec/Broadcom Endpoint Protection for Linux: exclusions are normally pushed from the SEP Manager console - a local agent usually can't add its own."),
    ("broadcom", "Symantec/Broadcom Endpoint Protection for Linux: exclusions are normally pushed from the SEP Manager console - a local agent usually can't add its own."),
    ("eset", "ESET Endpoint Antivirus for Linux: check /etc/opt/eset/esets/esets.cfg's on-access exclusion list, or the ESET PROTECT console if centrally managed."),
    ("kaspersky", "Kaspersky Endpoint Security for Linux: exclusions are set via the kesl-control CLI, or the centralized management console if enrolled."),
)

# (display name, systemd unit to check, process names to look for under
# /proc, representative install paths) - one entry per known Linux AV/EDR
# agent. A hit on ANY of the three signals counts as "installed", same
# spirit as Windows' SecurityCenter2 not distinguishing "installed" from
# "currently scanning". Curated from each vendor's own Linux agent
# documentation - deliberately not exhaustive, see module docstring.
_LINUX_AV_SIGNATURES: tuple[tuple[str, str, tuple[str, ...], tuple[str, ...]], ...] = (
    ("ClamAV", "clamav-daemon.service", ("clamd", "freshclam"), ("/usr/sbin/clamd", "/usr/bin/clamscan")),
    ("Microsoft Defender for Endpoint", "mdatp.service", ("wdavdaemon", "mdatp"), ("/opt/microsoft/mdatp/sbin/wdavdaemon",)),
    ("CrowdStrike Falcon", "falcon-sensor.service", ("falcon-sensor",), ("/opt/CrowdStrike/falcon-sensor",)),
    ("SentinelOne", "sentinelone.service", ("sentinelctl", "sentineld"), ("/opt/sentinelone/bin/sentinelctl",)),
    ("Sophos", "sophos-spl.service", ("sophos_threat_detector", "savdctl"), ("/opt/sophos-spl",)),
    ("Trend Micro Deep Security Agent", "ds_agent.service", ("ds_agent",), ("/opt/ds_agent/ds_agent",)),
    ("Bitdefender GravityZone", "bd.service", ("bdsec", "epag", "bdredline"), ("/opt/bitdefender-security-tools",)),
    ("Symantec/Broadcom Endpoint Protection", "sisidsdaemon.service", ("rtvscand", "symcfgd"), ("/opt/Symantec/symantec_antivirus",)),
    ("ESET Endpoint Antivirus", "esets.service", ("esets_daemon",), ("/opt/eset/esets/sbin/esets_daemon",)),
    ("Kaspersky Endpoint Security", "kesl.service", ("kesl",), ("/opt/kaspersky/kesl",)),
)


def looks_like_defender(name: str) -> bool:
    lowered = name.lower()
    return "defender" in lowered


def guidance_for(name: str) -> str:
    """Best-effort pointer to where this product's own exclusion settings
    usually live. Checks Linux-specific phrasing first, then falls back to
    the generic Windows-console-oriented table, then a generic, honest
    "consult the vendor" message for anything in neither table rather than
    guessing.
    """
    lowered = name.lower()
    for pattern, hint in _LINUX_VENDOR_GUIDANCE:
        if pattern in lowered:
            return hint
    for pattern, hint in _VENDOR_GUIDANCE:
        if pattern in lowered:
            return hint
    return f"No specific guidance available for {name} - check its settings for a scan exclusion/exception list."


def _systemd_unit_installed(unit: str) -> bool:
    """True if a unit file named `unit` exists on this system, regardless
    of whether it's currently active/enabled - closer to "is this product
    registered" than "is it running right now", matching the same spirit
    as the Windows SecurityCenter2 read. Silently returns False (not an
    error) when systemctl itself isn't present at all - some minimal/non-
    systemd environments (Alpine, certain containers) genuinely have no
    systemd, same "detection just won't find anything there" tradeoff as
    everything else in this table.
    """
    try:
        result = subprocess.run(
            ["systemctl", "list-unit-files", unit, "--no-legend"],
            capture_output=True, text=True, timeout=10,
        )
    except (FileNotFoundError, subprocess.TimeoutExpired):
        return False
    return bool((result.stdout or "").strip())


def _linux_process_names(proc_root: str | Path = "/proc") -> set[str]:
    """Lowercased set of every running process's comm name, read directly
    from /proc rather than shelling out to `ps` - pure filesystem access,
    no extra binary dependency, and works identically regardless of which
    `ps` variant (procps vs busybox vs toybox) happens to be installed.
    `proc_root` is overridable purely so tests can point this at a fake
    directory structure instead of the real /proc.
    """
    names: set[str] = set()
    root = Path(proc_root)
    if not root.is_dir():
        return names
    for entry in root.iterdir():
        if not entry.name.isdigit():
            continue
        try:
            names.add((entry / "comm").read_text(encoding="utf-8", errors="ignore").strip().lower())
        except OSError:
            continue
    return names


def _detect_linux() -> list[AvProduct]:
    """Checks every entry in _LINUX_AV_SIGNATURES against three signals
    (systemd unit file, running process, install path) and reports a hit
    on any match, deduplicated by display name in table order. See the
    module docstring for why this is a weaker guarantee than the Windows
    SecurityCenter2 path - an empty result means "nothing on the known
    list was found", not "nothing is installed".
    """
    running = _linux_process_names()
    products: list[AvProduct] = []
    for name, unit, process_names, install_paths in _LINUX_AV_SIGNATURES:
        hit = (
            _systemd_unit_installed(unit)
            or any(p in running for p in process_names)
            or any(Path(p).exists() for p in install_paths)
        )
        if hit:
            products.append(AvProduct(name=name))
    return products


def _detect_windows() -> list[AvProduct]:
    # -ErrorAction Stop turns "namespace doesn't exist" (Windows Server, or
    # the Security Center service disabled) into a catchable terminating
    # error instead of a silent empty result mixed in with stderr noise -
    # caught below and mapped to a clean empty list, same "not an error"
    # treatment as genuinely finding nothing installed.
    script = (
        "$ErrorActionPreference = 'Stop'; "
        "try { "
        "Get-CimInstance -Namespace root/SecurityCenter2 -ClassName AntiVirusProduct "
        "| Select-Object -ExpandProperty displayName | ConvertTo-Json -Compress "
        "} catch { '[]' }"
    )

    try:
        result = subprocess.run(
            ["powershell.exe", "-NoProfile", "-NonInteractive", "-Command", script],
            capture_output=True, text=True, timeout=30,
        )
    except FileNotFoundError as exc:
        raise AvDetectionError("powershell.exe was not found on PATH.") from exc
    except subprocess.TimeoutExpired as exc:
        raise AvDetectionError("Timed out querying installed antivirus products (30s).") from exc

    raw = (result.stdout or "").strip()
    if not raw:
        return []

    try:
        parsed = json.loads(raw)
    except json.JSONDecodeError:
        logger.warning("Could not parse antivirus product list from PowerShell output: %r", raw)
        return []

    # ConvertTo-Json returns a bare string for exactly one result, a real
    # array for zero or multiple - normalize to a list either way.
    if isinstance(parsed, str):
        names = [parsed]
    elif isinstance(parsed, list):
        names = [n for n in parsed if isinstance(n, str)]
    else:
        names = []

    seen: set[str] = set()
    products: list[AvProduct] = []
    for name in names:
        if name and name not in seen:
            seen.add(name)
            products.append(AvProduct(name=name))
    return products


def detect_av_products() -> list[AvProduct]:
    """Returns every antivirus/EDR product this machine's platform-specific
    detection mechanism found, deduplicated by name. An empty list is a
    normal, valid outcome on either supported platform (see module
    docstring for what "empty" means on each) - NOT an error. Raises
    AvDetectionError only for hard failures (Windows query couldn't run at
    all, or a platform other than Windows/Linux with no detection
    mechanism implemented yet).
    """
    system = platform.system()
    if system == "Windows":
        return _detect_windows()
    if system == "Linux":
        return _detect_linux()
    raise AvDetectionError(
        f"Antivirus detection isn't implemented for this platform ({system or 'unknown'}) yet - "
        "only Windows and Linux are currently supported."
    )

File: ingot/crates/ingot-core/src/av_detect.rs
//! Antivirus / EDR detection + a Windows-Defender scan-exclusion helper -
//! port of `binsifter.core.av_detect` and the (since-removed from Winnow)
//! `binsifter.core.defender`.
//!
//! Real-time protection races a bulk malware scan: a sample can be
//! quarantined between the moment Ingot enumerates it and the moment a
//! worker opens it, surfacing as an I/O error mid-scan. [`detect_av_products`]
//! answers "what AV is actually here", [`guidance_for`] points at that
//! vendor's own exclusion settings, and (Windows only, opt-in, one explicit
//! click) [`add_defender_exclusion`] adds a folder to Defender's exclusion
//! list via an elevated `Add-MpPreference`.
//!
//! Detection is two genuinely different implementations, not one with a
//! branch: Windows reads the `root/SecurityCenter2` WMI `AntiVirusProduct`
//! class (what Windows Security itself uses; no admin needed to read it);
//! Linux checks a curated table of systemd units / `/proc` process names /
//! install paths. macOS is not implemented. An empty result on either
//! platform is a valid outcome, never an error.

use std::path::Path;
use std::process::Command;

use base64::Engine;
use serde::Serialize;

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct AvProduct {
    pub name: String,
    /// Best-effort pointer to where this product's own exclusion list lives.
    pub guidance: String,
    /// True if this looks like Microsoft Defender (the one product the
    /// exclusion button can automate).
    pub is_defender: bool,
}

#[derive(Debug, thiserror::Error)]
pub enum AvError {
    #[error("{0}")]
    Detect(String),
    #[error("{0}")]
    Exclude(String),
}

pub fn looks_like_defender(name: &str) -> bool {
    name.to_ascii_lowercase().contains("defender")
}

/// (substring match, guidance). Windows-console-oriented; the Linux table
/// below is checked first.
const VENDOR_GUIDANCE: &[(&str, &str)] = &[
    ("windows defender", "Use the exclusion button below - Ingot can add this automatically."),
    ("microsoft defender", "Use the exclusion button below - Ingot can add this automatically."),
    ("mcafee", "McAfee: Endpoint Security console > Threat Prevention > Exclusions (or pushed from ePO if centrally managed)."),
    ("eset", "ESET: open the product > Setup > Detection Engine (or Antivirus) > Exclusions."),
    ("sophos", "Sophos: Sophos Endpoint > Exclusions, or Sophos Central > policy Exclusions if centrally managed."),
    ("symantec", "Symantec/Broadcom Endpoint Protection: exclusions are normally pushed from the SEP Manager console - a local client usually can't add its own."),
    ("broadcom", "Broadcom Endpoint Protection: exclusions are normally pushed from the SEP Manager console - a local client usually can't add its own."),
    ("crowdstrike", "CrowdStrike Falcon: exclusions are managed centrally from the Falcon console by design - there is no supported local self-exclusion."),
    ("sentinelone", "SentinelOne: exclusions are managed centrally from the Management Console by design - there is no supported local self-exclusion."),
    ("bitdefender", "Bitdefender: Protection > Antivirus > Settings > Manage Exceptions."),
    ("kaspersky", "Kaspersky: Settings > Threats and Exclusions > Manage Exclusions."),
    ("trend micro", "Trend Micro: open the console/agent's Scan Exclusion List settings."),
    ("malwarebytes", "Malwarebytes: Settings > Exclusions > Add Exclusion."),
    ("avast", "Avast: Menu > Settings > General > Exceptions."),
    ("avg", "AVG: Menu > Settings > General > Exceptions."),
    ("webroot", "Webroot: PC Security > Identity & Privacy Shields > Application/Exclusion list."),
    ("f-secure", "F-Secure: Settings > find the exclusion/exception list for real-time scanning."),
    ("norton", "Norton: Settings > Antivirus > Scans and Risks > Exclusions/Low Risks."),
];

/// Checked before [`VENDOR_GUIDANCE`] - a vendor's Linux agent usually
/// excludes via a config file / CLI, not the Windows GUI wording.
const LINUX_VENDOR_GUIDANCE: &[(&str, &str)] = &[
    ("clamav", "ClamAV: add an ExcludePath directive to /etc/clamav/clamd.conf, then restart the clamav-daemon service."),
    ("defender for endpoint", "Microsoft Defender for Endpoint on Linux: run `mdatp exclusion path add --path <folder>` - the button below only automates Windows Defender, not the Linux agent."),
    ("sophos", "Sophos for Linux: check your Sophos Central policy first if centrally managed; otherwise exclusions are set via the on-box config."),
    ("bitdefender", "Bitdefender GravityZone (Linux): exclusions are normally pushed from the GravityZone console, not set locally."),
    ("trend micro", "Trend Micro Deep Security Agent: exclusions are configured from the Deep Security Manager console."),
    ("symantec", "Symantec/Broadcom Endpoint Protection for Linux: exclusions are pushed from the SEP Manager console."),
    ("broadcom", "Symantec/Broadcom Endpoint Protection for Linux: exclusions are pushed from the SEP Manager console."),
    ("eset", "ESET Endpoint Antivirus for Linux: check /etc/opt/eset/esets/esets.cfg's exclusion list, or the ESET PROTECT console."),
    ("kaspersky", "Kaspersky Endpoint Security for Linux: exclusions are set via the kesl-control CLI, or the management console."),
];

pub fn guidance_for(name: &str) -> String {
    let lowered = name.to_ascii_lowercase();
    for (pat, hint) in LINUX_VENDOR_GUIDANCE {
        if lowered.contains(pat) {
            return (*hint).to_string();
        }
    }
    for (pat, hint) in VENDOR_GUIDANCE {
        if lowered.contains(pat) {
            return (*hint).to_string();
        }
    }
    format!(
        "No specific guidance for {name} - check its settings for a scan exclusion/exception list."
    )
}

fn product(name: &str) -> AvProduct {
    AvProduct {
        name: name.to_string(),
        guidance: guidance_for(name),
        is_defender: looks_like_defender(name),
    }
}

/// Every AV/EDR product this platform's detection found, deduplicated by
/// name. An empty list is valid (see module docs). Errors only on a hard
/// failure (Windows query can't run) or an unsupported platform.
pub fn detect_av_products() -> Result<Vec<AvProduct>, AvError> {
    #[cfg(target_os = "windows")]
    {
        detect_windows()
    }
    #[cfg(target_os = "linux")]
    {
        Ok(detect_linux())
    }
    #[cfg(not(any(target_os = "windows", target_os = "linux")))]
    {
        Err(AvError::Detect(
            "Antivirus detection isn't implemented for this platform yet - only Windows and Linux."
                .to_string(),
        ))
    }
}

#[cfg(target_os = "windows")]
fn detect_windows() -> Result<Vec<AvProduct>, AvError> {
    // -ErrorAction Stop turns "namespace missing" (Windows Server, Security
    // Center disabled) into a catchable error mapped to an empty list.
    let script = "$ErrorActionPreference='Stop'; try { \
        Get-CimInstance -Namespace root/SecurityCenter2 -ClassName AntiVirusProduct \
        | Select-Object -ExpandProperty displayName | ConvertTo-Json -Compress \
        } catch { '[]' }";

    let out = Command::new("powershell.exe")
        .args(["-NoProfile", "-NonInteractive", "-Command", script])
        .output()
        .map_err(|e| AvError::Detect(format!("could not run powershell.exe: {e}")))?;

    let raw = String::from_utf8_lossy(&out.stdout);
    let raw = raw.trim();
    if raw.is_empty() {
        return Ok(Vec::new());
    }
    // ConvertTo-Json: a bare string for one result, an array otherwise.
    let names: Vec<String> = match serde_json::from_str::<serde_json::Value>(raw) {
        Ok(serde_json::Value::String(s)) => vec![s],
        Ok(serde_json::Value::Array(a)) => a
            .into_iter()
            .filter_map(|v| v.as_str().map(str::to_string))
            .collect(),
        _ => return Ok(Vec::new()),
    };

    let mut seen = std::collections::HashSet::new();
    Ok(names
        .into_iter()
        .filter(|n| !n.is_empty() && seen.insert(n.clone()))
        .map(|n| product(&n))
        .collect())
}

/// (display name, systemd unit, process names in /proc, representative
/// install paths). A hit on ANY signal counts as installed.
#[cfg(target_os = "linux")]
const LINUX_AV_SIGNATURES: &[(&str, &str, &[&str], &[&str])] = &[
    (
        "ClamAV",
        "clamav-daemon.service",
        &["clamd", "freshclam"],
        &["/usr/sbin/clamd", "/usr/bin/clamscan"],
    ),
    (
        "Microsoft Defender for Endpoint",
        "mdatp.service",
        &["wdavdaemon", "mdatp"],
        &["/opt/microsoft/mdatp/sbin/wdavdaemon"],
    ),
    (
        "CrowdStrike Falcon",
        "falcon-sensor.service",
        &["falcon-sensor"],
        &["/opt/CrowdStrike/falcon-sensor"],
    ),
    (
        "SentinelOne",
        "sentinelone.service",
        &["sentinelctl", "sentineld"],
        &["/opt/sentinelone/bin/sentinelctl"],
    ),
    (
        "Sophos",
        "sophos-spl.service",
        &["sophos_threat_detector", "savdctl"],
        &["/opt/sophos-spl"],
    ),
    (
        "Trend Micro Deep Security Agent",
        "ds_agent.service",
        &["ds_agent"],
        &["/opt/ds_agent/ds_agent"],
    ),
    (
        "Bitdefender GravityZone",
        "bd.service",
        &["bdsec", "epag", "bdredline"],
        &["/opt/bitdefender-security-tools"],
    ),
    (
        "Symantec/Broadcom Endpoint Protection",
        "sisidsdaemon.service",
        &["rtvscand", "symcfgd"],
        &["/opt/Symantec/symantec_antivirus"],
    ),
    (
        "ESET Endpoint Antivirus",
        "esets.service",
        &["esets_daemon"],
        &["/opt/eset/esets/sbin/esets_daemon"],
    ),
    (
        "Kaspersky Endpoint Security",
        "kesl.service",
        &["kesl"],
        &["/opt/kaspersky/kesl"],
    ),
];

#[cfg(target_os = "linux")]
fn detect_linux() -> Vec<AvProduct> {
    let running = linux_process_names();
    LINUX_AV_SIGNATURES
        .iter()
        .filter(|(_, unit, procs, paths)| {
            systemd_unit_installed(unit)
                || procs.iter().any(|p| running.contains(*p))
                || paths.iter().any(|p| Path::new(p).exists())
        })
        .map(|(name, ..)| product(name))
        .collect()
}

#[cfg(target_os = "linux")]
fn systemd_unit_installed(unit: &str) -> bool {
    Command::new("systemctl")
        .args(["list-unit-files", unit, "--no-legend"])
        .output()
        .map(|o| !String::from_utf8_lossy(&o.stdout).trim().is_empty())
        .unwrap_or(false)
}

#[cfg(target_os = "linux")]
fn linux_process_names() -> std::collections::HashSet<String> {
    let mut names = std::collections::HashSet::new();
    let Ok(entries) = std::fs::read_dir("/proc") else {
        return names;
    };
    for entry in entries.flatten() {
        if entry
            .file_name()
            .to_string_lossy()
            .chars()
            .all(|c| c.is_ascii_digit())
        {
            if let Ok(comm) = std::fs::read_to_string(entry.path().join("comm")) {
                names.insert(comm.trim().to_string());
            }
        }
    }
    names
}

/// Add `folder` to Windows Defender's scan-exclusion list. Windows only,
/// opt-in (one explicit click + confirmation on the caller's side). Spawns a
/// **separate** elevated PowerShell via `Start-Process -Verb RunAs` so
/// Windows' own UAC dialog does the elevation - Ingot never gains admin.
pub fn add_defender_exclusion(folder: &Path) -> Result<(), AvError> {
    if !cfg!(target_os = "windows") {
        return Err(AvError::Exclude(
            "Windows Defender exclusions are a Windows-only feature.".to_string(),
        ));
    }
    // best-effort: the folder existing means the exclusion is immediately
    // visible in Windows Security's own UI. Add-MpPreference doesn't require it.
    let _ = std::fs::create_dir_all(folder);

    let escaped = folder.to_string_lossy().replace('\'', "''");
    let inner =
        format!("try {{ Add-MpPreference -ExclusionPath '{escaped}'; exit 0 }} catch {{ exit 1 }}");
    // -EncodedCommand wants base64 of UTF-16LE - avoids nested-quote hell.
    let mut utf16 = Vec::new();
    for u in inner.encode_utf16() {
        utf16.extend_from_slice(&u.to_le_bytes());
    }
    let encoded = base64::engine::general_purpose::STANDARD.encode(&utf16);

    // Start-Process -Verb RunAs *throws* (rather than returning a bad exit
    // code) when UAC is declined - catch it and map to 1223 (ERROR_CANCELLED).
    let outer = format!(
        "$ErrorActionPreference='Stop'; try {{ \
         $p = Start-Process powershell.exe -Verb RunAs -Wait -PassThru \
         -ArgumentList '-NoProfile -NonInteractive -EncodedCommand {encoded}'; \
         exit $p.ExitCode }} catch {{ exit 1223 }}"
    );

    let out = Command::new("powershell.exe")
        .args(["-NoProfile", "-Command", &outer])
        .output()
        .map_err(|e| AvError::Exclude(format!("could not run powershell.exe: {e}")))?;

    match out.status.code() {
        Some(0) => Ok(()),
        Some(1223) => Err(AvError::Exclude(
            "UAC elevation was declined (or the elevated process could not start) - no changes made."
                .to_string(),
        )),
        _ => Err(AvError::Exclude(
            "Add-MpPreference failed. Usually this means Defender isn't the active antivirus, \
             its real-time protection is off, or Tamper Protection is blocking preference changes."
                .to_string(),
        )),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn defender_recognised() {
        assert!(looks_like_defender("Windows Defender"));
        assert!(looks_like_defender("Microsoft Defender Antivirus"));
        assert!(!looks_like_defender("ESET Endpoint Antivirus"));
    }

    #[test]
    fn guidance_prefers_linux_wording() {
        assert!(guidance_for("ClamAV").contains("clamd.conf"));
        assert!(guidance_for("ESET Endpoint Antivirus for Linux").contains("esets.cfg"));
        // unknown product -> honest fallback, not a guess
        assert!(guidance_for("Frobozz AV").contains("No specific guidance"));
    }

    #[test]
    fn exclusion_is_windows_only_off_windows() {
        if !cfg!(target_os = "windows") {
            assert!(add_defender_exclusion(Path::new("/tmp/x")).is_err());
        }
    }
}
